File: app/src/main/java/org/ole/planet/myplanet/base/BaseResourceFragment.java
package org.ole.planet.myplanet.base;

import android.app.ProgressDialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.net.wifi.WifiManager;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.Fragment;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import org.ole.planet.myplanet.MainApplication;
import org.ole.planet.myplanet.R;
import org.ole.planet.myplanet.callback.OnHomeItemClickListener;
import org.ole.planet.myplanet.datamanager.DatabaseService;
import org.ole.planet.myplanet.datamanager.Service;
import org.ole.planet.myplanet.model.Download;
import org.ole.planet.myplanet.model.RealmMyCourse;
import org.ole.planet.myplanet.model.RealmMyLibrary;
import org.ole.planet.myplanet.model.RealmRemovedLog;
import org.ole.planet.myplanet.model.RealmStepExam;
import org.ole.planet.myplanet.model.RealmSubmission;
import org.ole.planet.myplanet.model.RealmTag;
import org.ole.planet.myplanet.model.RealmUserModel;
import org.ole.planet.myplanet.service.UserProfileDbHandler;
import org.ole.planet.myplanet.ui.submission.AdapterMySubmission;
import org.ole.planet.myplanet.ui.sync.SyncActivity;
import org.ole.planet.myplanet.utilities.CheckboxListView;
import org.ole.planet.myplanet.utilities.DialogUtils;
import org.ole.planet.myplanet.utilities.DownloadUtils;
import org.ole.planet.myplanet.utilities.Utilities;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import io.realm.Realm;
import io.realm.RealmObject;
import io.realm.RealmResults;

import static android.content.Context.MODE_PRIVATE;
import static org.ole.planet.myplanet.ui.dashboard.DashboardActivity.MESSAGE_PROGRESS;

public abstract class BaseResourceFragment extends Fragment {
    public static SharedPreferences settings;
    public static String auth = "";
    static ProgressDialog prgDialog;
    public OnHomeItemClickListener homeItemClickListener;
    public RealmUserModel model;
    public Realm mRealm;
    public UserProfileDbHandler profileDbHandler;
    //    ArrayList<Integer> selectedItemsList = new ArrayList<>();
    CheckboxListView lv;
    View convertView;
    BroadcastReceiver receiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            Utilities.log("Broad cast received");
            showDownloadDialog(getLibraryList(new DatabaseService(context).getRealmInstance()));
        }
    };

    BroadcastReceiver stateReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            new AlertDialog.Builder(getActivity())
                    .setMessage("Do you want to stay online?")
                    .setPositiveButton("Yes", null)
                    .setNegativeButton("No", (dialogInterface, i) -> {
                        WifiManager wifi = (WifiManager) MainApplication.context.getSystemService(Context.WIFI_SERVICE);
                        if (wifi != null)
                            wifi.setWifiEnabled(false);
                    })
                    .show();
        }
    };
    private BroadcastReceiver broadcastReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (intent.getAction().equals(MESSAGE_PROGRESS) && prgDialog != null) {
                Download download = intent.getParcelableExtra("download");
                if (!download.isFailed()) {
                    setProgress(download);
                } else {
                    DialogUtils.showError(prgDialog, download.getMessage());
                }
            }
        }
    };

    protected void showDownloadDialog(final List<RealmMyLibrary> db_myLibrary) {
        new Service(MainApplication.context).isPlanetAvailable(new Service.PlanetAvailableListener() {
            @Override
            public void isAvailable() {
                if (!db_myLibrary.isEmpty()) {
                    LayoutInflater inflater = getLayoutInflater();
                    convertView = inflater.inflate(R.layout.my_library_alertdialog, null);
                    AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(getActivity());
                    alertDialogBuilder.setView(convertView).setTitle(R.string.download_suggestion);
                    alertDialogBuilder.setPositiveButton(R.string.download_selected, (dialogInterface, i) -> startDownload(DownloadUtils.downloadFiles(db_myLibrary, lv.getSelectedItemsList(), settings))).setNeutralButton(R.string.download_all, (dialogInterface, i) -> startDownload(DownloadUtils.downloadAllFiles(db_myLibrary, settings))).setNegativeButton(R.string.txt_cancel, null);
                    AlertDialog alertDialog = alertDialogBuilder.create();
                    createListView(db_myLibrary, alertDialog);
                    alertDialog.show();
                    (alertDialog).getButton(AlertDialog.BUTTON_POSITIVE).setEnabled(lv.getSelectedItemsList().size() > 0);
                }else{
                    Utilities.toast(requireContext(), "No resources to download");
                }
            }

            @Override
            public void notAvailable() {
                Utilities.toast(requireContext(), "Planet not available");

                Utilities.log("Planet not available");
            }
        });
    }

    public void showPendingSurveyDialog() {
        model = new UserProfileDbHandler(getActivity()).getUserModel();
        List<RealmSubmission> list = mRealm.where(RealmSubmission.class).equalTo("userId", model.getId()).equalTo("status", "pending").equalTo("type", "survey").findAll();
        if (list.size() == 0) {
            return;
        }
        HashMap<String, RealmStepExam> exams = RealmSubmission.getExamMap(mRealm, list);
        ArrayAdapter arrayAdapter = new ArrayAdapter(getActivity(), android.R.layout.simple_list_item_1, list) {
            @NonNull
            @Override
            public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
                if (convertView == null)
                    convertView = LayoutInflater.from(getActivity()).inflate(android.R.layout.simple_list_item_1, null);
                if (exams.containsKey(((RealmSubmission) getItem(position)).getParentId()))
                    ((TextView) convertView).setText(exams.get(list.get(position).getParentId()).getName());
                else {
                    ((TextView) convertView).setText("N/A");
                }
                return convertView;
            }
        };
        new AlertDialog.Builder(getActivity()).setTitle("Pending Surveys").setAdapter(arrayAdapter, (dialogInterface, i) -> AdapterMySubmission.openSurvey(homeItemClickListener, list.get(i).getId(), true)).setPositiveButton("Dismiss", null).show();
    }

    public void startDownload(ArrayList urls) {
        new Service(getActivity()).isPlanetAvailable(new Service.PlanetAvailableListener() {
            @Override
            public void isAvailable() {
                if (!urls.isEmpty()) {
                    prgDialog.show();
                    Utilities.openDownloadService(getActivity(), urls);
                }
            }

            @Override
            public void notAvailable() {
                Utilities.toast(getActivity(), "Device not connected to planet.");
            }
        });

    }

    public void setProgress(Download download) {
        prgDialog.setProgress(download.getProgress());
        if (!TextUtils.isEmpty(download.getFileName())) {
            prgDialog.setTitle(download.getFileName());
        }
        if (download.isCompleteAll()) {
            DialogUtils.showError(prgDialog, "All files downloaded successfully");
            onDownloadComplete();
        }
    }


    public void onDownloadComplete() {
    }

    public void createListView(List<RealmMyLibrary> db_myLibrary, AlertDialog alertDialog) {
        lv = convertView.findViewById(R.id.alertDialog_listView);
        ArrayList<String> names = new ArrayList<>();
        for (int i = 0; i < db_myLibrary.size(); i++) {
            names.add(db_myLibrary.get(i).getTitle());
        }
        ArrayAdapter<String> adapter = new ArrayAdapter<String>(getActivity().getBaseContext(), R.layout.rowlayout, R.id.checkBoxRowLayout, names);
        lv.setChoiceMode(ListView.CHOICE_MODE_MULTIPLE);
        lv.setCheckChangeListener(() -> {
            (alertDialog).getButton(AlertDialog.BUTTON_POSITIVE).setEnabled(lv.getSelectedItemsList().size() > 0);
        });
        lv.setAdapter(adapter);
    }


    private void registerReceiver() {
        LocalBroadcastManager bManager = LocalBroadcastManager.getInstance(getActivity());
        IntentFilter intentFilter = new IntentFilter();
        intentFilter.addAction(MESSAGE_PROGRESS);
        bManager.registerReceiver(broadcastReceiver, intentFilter);
        IntentFilter intentFilter2 = new IntentFilter();
        intentFilter2.addAction("ACTION_NETWORK_CHANGED");
        LocalBroadcastManager.getInstance(MainApplication.context).registerReceiver(receiver, intentFilter2);
        IntentFilter intentFilter3 = new IntentFilter();
        intentFilter3.addAction("SHOW_WIFI_ALERT");
        LocalBroadcastManager.getInstance(MainApplication.context).registerReceiver(stateReceiver, intentFilter3);
    }

    public List<RealmMyLibrary> getLibraryList(Realm mRealm) {
        return getLibraryList(mRealm, settings.getString("userId", "--"));
    }

    public List<RealmMyLibrary> getLibraryList(Realm mRealm, String userId) {
        RealmResults<RealmMyLibrary> l = mRealm.where(RealmMyLibrary.class).equalTo("isPrivate", false).findAll();
        List<RealmMyLibrary> libList = new ArrayList<>();
        List<RealmMyLibrary> libraries = getLibraries(l);
        for (RealmMyLibrary item : libraries) {
            if (item.getUserId().contains(userId)) {
                libList.add(item);
            }
        }
        return libList;
    }

    private List<RealmMyLibrary> getLibraries(RealmResults<RealmMyLibrary> l) {
        List<RealmMyLibrary> libraries = new ArrayList<>();
        for (RealmMyLibrary lib : l) {
            if (lib.needToUpdate()) {
                libraries.add(lib);
            }
        }
        return libraries;
    }


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mRealm = new DatabaseService(getActivity()).getRealmInstance();
        prgDialog = DialogUtils.getProgressDialog(getActivity());
        settings = getActivity().getSharedPreferences(SyncActivity.PREFS_NAME, MODE_PRIVATE);

    }

    @Override
    public void onPause() {
        super.onPause();
        LocalBroadcastManager.getInstance(getActivity()).unregisterReceiver(receiver);
        LocalBroadcastManager.getInstance(getActivity()).unregisterReceiver(broadcastReceiver);
        LocalBroadcastManager.getInstance(getActivity()).unregisterReceiver(stateReceiver);

    }

    public void removeFromShelf(RealmObject object) {
        if (object instanceof RealmMyLibrary) {
            RealmMyLibrary myObject = mRealm.where(RealmMyLibrary.class).equalTo("resourceId", ((RealmMyLibrary) object).getResource_id()).findFirst();
            myObject.removeUserId(model.getId());
            RealmRemovedLog.onRemove(mRealm, "resources", model.getId(), ((RealmMyLibrary) object).getResource_id());
            Utilities.toast(getActivity(), "Removed from myLibrary");
        } else {
            RealmMyCourse myObject = RealmMyCourse.getMyCourse(mRealm, ((RealmMyCourse) object).getCourseId());
            myObject.removeUserId(model.getId());
            RealmRemovedLog.onRemove(mRealm, "courses", model.getId(), ((RealmMyCourse) object).getCourseId());
            Utilities.toast(getActivity(), "Removed from myCourse");
        }
    }


    @Override
    public void onResume() {
        super.onResume();
        registerReceiver();

    }


    public void showTagText(List<RealmTag> list, TextView tvSelected) {
        StringBuilder selected = new StringBuilder("Selected : ");
        for (RealmTag tags :
                list) {
            selected.append(tags.getName()).append(",");
        }
        tvSelected.setText(selected.subSequence(0, selected.length() - 1));
    }

}

File: app/src/main/java/org/ole/planet/myplanet/base/BaseResourceFragment.kt
package org.ole.planet.myplanet.base

import android.content.BroadcastReceiver
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import dagger.hilt.android.AndroidEntryPoint
import io.realm.RealmObject
import javax.inject.Inject
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import org.ole.planet.myplanet.R
import org.ole.planet.myplanet.callback.OnHomeItemClickListener
import org.ole.planet.myplanet.model.Download
import org.ole.planet.myplanet.model.RealmMyCourse
import org.ole.planet.myplanet.model.RealmMyLibrary
import org.ole.planet.myplanet.model.RealmTag
import org.ole.planet.myplanet.model.RealmUser
import org.ole.planet.myplanet.repository.CoursesRepository
import org.ole.planet.myplanet.repository.ResourcesRepository
import org.ole.planet.myplanet.repository.SubmissionsRepository
import org.ole.planet.myplanet.repository.SurveysRepository
import org.ole.planet.myplanet.repository.UserRepository
import org.ole.planet.myplanet.services.BroadcastService
import org.ole.planet.myplanet.services.DownloadService
import org.ole.planet.myplanet.services.SharedPrefManager
import org.ole.planet.myplanet.services.UserSessionManager
import org.ole.planet.myplanet.ui.components.CheckboxAdapter
import org.ole.planet.myplanet.ui.dashboard.DashboardActivity
import org.ole.planet.myplanet.utils.DialogUtils
import org.ole.planet.myplanet.utils.DialogUtils.getProgressDialog
import org.ole.planet.myplanet.utils.DialogUtils.showError
import org.ole.planet.myplanet.utils.Utilities

@AndroidEntryPoint
abstract class BaseResourceFragment : Fragment() {
    var homeItemClickListener: OnHomeItemClickListener? = null
    var model: RealmUser? = null
    var lv: RecyclerView? = null
    var convertView: View? = null
    internal lateinit var prgDialog: DialogUtils.CustomProgressDialog
    @Inject
    lateinit var userRepository: UserRepository
    @Inject
    lateinit var resourcesRepository: ResourcesRepository
    @Inject
    lateinit var coursesRepository: CoursesRepository
    @Inject
    lateinit var submissionsRepository: SubmissionsRepository
    @Inject
    lateinit var surveysRepository: SurveysRepository
    @Inject
    lateinit var profileDbHandler: UserSessionManager
    @Inject
    lateinit var sharedPrefManager: SharedPrefManager
    @Inject
    lateinit var broadcastService: BroadcastService
    private var resourceNotFoundDialog: AlertDialog? = null
    private var downloadSuggestionDialog: AlertDialog? = null

    private fun isFragmentActive(): Boolean {
        return isAdded && activity != null &&
            !requireActivity().isFinishing && !requireActivity().isDestroyed
    }

    internal fun showProgressDialog() {
        viewLifecycleOwner.lifecycleScope.launch {
            if (isFragmentActive()) {
                prgDialog.setIndeterminateMode(true)
                prgDialog.show()
            }
        }
    }

    internal fun showNotConnectedToast() {
        if (isFragmentActive()) {
            Utilities.toast(requireActivity(),
                getString(R.string.device_not_connected_to_planet))
        }
    }

    private var receiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val pendingResult = goAsync()
            this@BaseResourceFragment.lifecycleScope.launch {
                try {
                    val list = resourcesRepository.getDownloadSuggestionList()
                    showDownloadDialog(list)
                } finally {
                    pendingResult.finish()
                }
            }
        }
    }
    private val pendingDownloadUrls = mutableSetOf<String>()

    protected fun trackDownloadUrls(urls: Collection<String>) {
        pendingDownloadUrls.clear()
        pendingDownloadUrls.addAll(urls)
    }

    private val broadcastReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (intent.action == DashboardActivity.MESSAGE_PROGRESS) {
                val download: Download? = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                    intent.getParcelableExtra("download", Download::class.java)
                } else {
                    @Suppress("DEPRECATION")
                    intent.getParcelableExtra("download")
                }
                if (download?.failed == false) {
                    if (pendingDownloadUrls.isNotEmpty()) {
                        val fileUrl = download.fileUrl
                        if (!fileUrl.isNullOrEmpty() && fileUrl in pendingDownloadUrls) {
                            if (download.progress == 100) {
                                pendingDownloadUrls.remove(fileUrl)
                                onSingleResourceDownloaded(fileUrl)
                            }
                            setProgress(download.apply { completeAll = pendingDownloadUrls.isEmpty() })
                        }
                    }
                } else {
                    pendingDownloadUrls.clear()
                    download?.message?.let { showError(prgDialog, it) } ?: prgDialog.dismiss()
                }
            }
        }
    }

    protected fun showDownloadDialog(dbMyLibrary: List<RealmMyLibrary?>) {
        if (!isAdded) return
        if (dbMyLibrary.isEmpty()) {
            return
        }

        activity?.let { fragmentActivity ->
            val inflater = fragmentActivity.layoutInflater
            val rootView = fragmentActivity.findViewById<ViewGroup>(android.R.id.content)
            convertView = inflater.inflate(R.layout.my_library_alertdialog, rootView, false)

            val alertDialogBuilder = AlertDialog.Builder(fragmentActivity, R.style.AlertDialogTheme)
            val titleView = TextView(requireContext()).apply {
                text = getString(R.string.download_suggestion)
                setPadding(48, 40, 48, 0)
                textSize = 18f
                maxLines = 5
                isSingleLine = false
                setTextColor(ContextCompat.getColor(requireContext(), R.color.daynight_textColor))
            }
            alertDialogBuilder.setView(convertView)
                .setCustomTitle(titleView)


                .setPositiveButton(R.string.download_selected) { _: DialogInterface?, _: Int ->
                    lifecycleScope.launch {
                        val selectedItemsList = (lv?.adapter as? CheckboxAdapter)?.selectedItemsList
                        selectedItemsList?.let {
                            addToLibrary(dbMyLibrary, ArrayList(it))
                            val selectedLibraries = it.mapNotNull { index -> dbMyLibrary.getOrNull(index) }
                            if (resourcesRepository.downloadResources(selectedLibraries)) {
                                val urls = selectedLibraries.mapNotNull { lib -> lib.resourceRemoteAddress }
                                trackDownloadUrls(urls)
                                showProgressDialog()
                            }
                        }
                    }
                }.setNeutralButton(R.string.download_all) { _: DialogInterface?, _: Int ->
                    lifecycleScope.launch {
                        addAllToLibrary(dbMyLibrary)
                        val filtered = dbMyLibrary.filterNotNull()
                        if (resourcesRepository.downloadResources(filtered)) {
                            val urls = filtered.mapNotNull { lib -> lib.resourceRemoteAddress }
                            trackDownloadUrls(urls)
                            showProgressDialog()
                        }
                    }
                }.setNegativeButton(R.string.txt_cancel, null)
            downloadSuggestionDialog?.dismiss()
            downloadSuggestionDialog = alertDialogBuilder.create()
            downloadSuggestionDialog?.let { dialog ->
                createListView(dbMyLibrary, dialog)
                dialog.setOnDismissListener {
                    downloadSuggestionDialog = null
                }
                dialog.show()
                dialog.getButton(AlertDialog.BUTTON_POSITIVE).isEnabled = ((lv?.adapter as? CheckboxAdapter)?.selectedItemsList?.size
                    ?: 0) > 0
            }
        }
    }

    private val resourceNotFoundReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            showResourceNotFoundDialog()
        }
    }

    private fun showResourceNotFoundDialog() {
        if (isAdded) {
            if (prgDialog.isShowing()) {
                prgDialog.dismiss()
            }

            if (resourceNotFoundDialog?.isShowing == true) {
                return
            }

            resourceNotFoundDialog = AlertDialog.Builder(requireContext(), R.style.AlertDialogTheme)
                .setTitle(R.string.resource_not_found)
                .setMessage(R.string.resource_not_found_message)
                .setNegativeButton(R.string.close) { dialog, _ ->
                    dialog.dismiss()
                }
                .create()
            resourceNotFoundDialog?.setOnDismissListener {
                resourceNotFoundDialog = null
            }

            resourceNotFoundDialog?.show()
        }
    }

    fun setProgress(download: Download) {
        prgDialog.setProgress(download.progress)
        if (!TextUtils.isEmpty(download.fileName)) {
            prgDialog.setTitle(download.fileName)
        }
        if (download.completeAll) {
            onDownloadComplete()
        }
    }

    protected open fun onSingleResourceDownloaded(url: String) {}

    open fun onDownloadComplete() {
        prgDialog.dismiss()

        if (sharedPrefManager.isAlternativeUrl()) {
            sharedPrefManager.setAlternativeUrl("")
            sharedPrefManager.setProcessedAlternativeUrl("")
            sharedPrefManager.setIsAlternativeUrl(false)
        }
    }

    fun createListView(dbMyLibrary: List<RealmMyLibrary?>, alertDialog: AlertDialog) {
        lv = convertView?.findViewById(R.id.alertDialog_listView)
        val names = dbMyLibrary.map { it?.title ?: "" }
        val adapter = CheckboxAdapter {
            alertDialog.getButton(AlertDialog.BUTTON_POSITIVE).isEnabled = ((lv?.adapter as? CheckboxAdapter)?.selectedItemsList?.size ?: 0) > 0
        }
        adapter.submitList(names)
        lv?.layoutManager = LinearLayoutManager(requireActivity().baseContext)
        lv?.adapter = adapter
    }

    private fun registerReceiver() {
        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                broadcastService.events.collect { intent ->
                    if (isActive) {
                        when (intent.action) {
                            DashboardActivity.MESSAGE_PROGRESS -> broadcastReceiver.onReceive(requireContext(), intent)
                            "ACTION_NETWORK_CHANGED" -> receiver.onReceive(requireContext(), intent)
                            DownloadService.RESOURCE_NOT_FOUND_ACTION -> resourceNotFoundReceiver.onReceive(requireContext(), intent)
                        }
                    }
                }
            }
        }
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        registerReceiver()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        prgDialog = getProgressDialog(requireActivity())
    }

    override fun onDetach() {
        super.onDetach()
        homeItemClickListener = null
    }

    fun removeFromShelf(`object`: RealmObject) {
        lifecycleScope.launch {
            val userId = profileDbHandler.getUserModel()?.id
            if (userId.isNullOrEmpty()) {
                return@launch
            }

            if (`object` is RealmMyLibrary) {
                val resourceId = `object`.resourceId
                if (resourceId != null) {
                    resourcesRepository.removeResourceFromShelf(resourceId, userId)
                    Utilities.toast(activity, getString(R.string.removed_from_mylibrary))
                }
            } else {
                val courseId = (`object` as RealmMyCourse).courseId
                if (courseId != null) {
                    coursesRepository.removeCourseFromShelf(courseId, userId)
                    Utilities.toast(activity, getString(R.string.removed_from_mycourse))
                }
            }
        }
    }

    fun showTagText(list: List<RealmTag>, tvSelected: TextView?) {
        val selected = list.joinToString(separator = ",", prefix = getString(R.string.selected)) { it.name.orEmpty() }
        tvSelected?.text = selected
    }

    fun addToLibrary(libraryItems: List<RealmMyLibrary?>, selectedItems: ArrayList<Int>) {
        lifecycleScope.launch {
            val userId = profileDbHandler.getUserModel()?.id ?: return@launch
            val resourceIds = selectedItems.mapNotNull { index ->
                libraryItems.getOrNull(index)?.resourceId
            }
            resourcesRepository.addResourcesToUserLibrary(resourceIds, userId)
                .onSuccess {
                    Utilities.toast(activity, getString(R.string.added_to_my_library))
                }
                .onFailure {
                    Utilities.toast(activity, getString(R.string.error, it.message))
                }
        }
    }

    fun addAllToLibrary(libraryItems: List<RealmMyLibrary?>) {
        lifecycleScope.launch {
            val user = profileDbHandler.getUserModel()
            val userId = user?.id ?: return@launch
            val validLibraryItems = libraryItems.filterNotNull()
            resourcesRepository.addAllResourcesToUserLibrary(validLibraryItems, userId)
                .onSuccess {
                    Utilities.toast(activity, getString(R.string.added_to_my_library))
                }
                .onFailure {
                    Utilities.toast(activity, getString(R.string.error, it.message))
                }
        }
    }

    override fun onDestroyView() {
        downloadSuggestionDialog?.dismiss()
        downloadSuggestionDialog = null
        resourceNotFoundDialog?.dismiss()
        resourceNotFoundDialog = null
        convertView = null
        super.onDestroyView()
    }
}
